### PR TITLE
initial support for registry publishing

### PR DESCRIPTION
**Pull request by tomerd**
motivation: support publishing to registry

changes:
* split SwiftPackageRegistryTool to multiple files as its growing large
* add "swift package-registry publish" command
* implement client side for registry requirments API
* add test

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

**Comment by tomerd** [comment]
macOS CI failure related to https://github.com/apple/swift-package-manager/pull/6103

**Comment by tomerd** [comment]
cc @yim-lee

**Comment by tomerd** [comment]
@swift-ci smoke test macOS

File: Sources/PackageRegistryTool/SwiftPackageRegistryTool+Publish.swift [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+                if let registry = configuration.registry(for: packageScopeAndName.scope) {
+                    return registry.url
+                }
+                if let registry = configuration.defaultRegistry {
+                    return registry.url
+                }
+                return .none
+            }()
+
+            guard let registryURL else {
```
Comment (MaxDesiatov) [inline Sources/PackageRegistryTool/SwiftPackageRegistryTool+Publish.swift:79]: this is apparently causing a failure on macOS CI nodes which have Swift 5.5
```suggestion
            guard let registryURL = registryURL else {
```

File: Sources/CoreCommands/Options.swift [hunk 2]
```diff
@@ -451,13 +491,27 @@ public extension Sanitizer {
             return
         }
 
-        throw StringError("valid sanitizers: \(Sanitizer.formattedValues)")
```
Comment (neonichu) [inline Sources/CoreCommands/Options.swift:454]: This seems like a useful error we're losing here

**Comment by tomerd** [comment]
@swift-ci smoke test

File: Sources/CoreCommands/Options.swift (hunk 2 above)
Comment (tomerd) [reply Sources/CoreCommands/Options.swift:454]: its actually not needed as it is part of the help string that is presented in this case

**Comment by tomerd** [comment]
@swift-ci smoke test

**Review by yim-lee — APPROVED** [review]
Looks reasonable, though the proposal is still in draft state and there is high possibility for registry APIs to change still.

File: Sources/PackageRegistryTool/SwiftPackageRegistryTool+Publish.swift [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            }
+
+            let registryClient = RegistryClient(
+                configuration: configuration,
+                fingerprintStorage: .none,
+                fingerprintCheckingMode: .strict,
+                authorizationProvider: authorizationProvider
+            )
+
+            let publishRequirements = try tsc_await { callback in
```
Comment (MaxDesiatov) [inline Sources/PackageRegistryTool/SwiftPackageRegistryTool+Publish.swift:98]: This now causes a warning on `main`: 
```
Initialization of immutable value 'publishRequirements' was never used; consider replacing with assignment to '_' or removing it
```

Comment (tomerd) [reply Sources/PackageRegistryTool/SwiftPackageRegistryTool+Publish.swift:98]: right. this is solved in the followup PR: https://github.com/apple/swift-package-manager/pull/6118

**Pull request merged**